keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,364 @@
+* Persistent inline volumes do not participate in the provision and deletion phases.
+* Persistent inline volumes expect provision and deletion to be handled outside of the driver.
+
+For ephemeral CSI drivers:
+* Ephemeral drivers will not receive provision/deletion API calls (since these stages are driven by a PV/PVC/StorageClass).
+* Ephemeral CSI drivers will have to delay or combine any provisioning/deprovisioning operation during different phase.
+
+#### Attach/detach
+CSI uses API object `storage.VolumeAttachment` to track and manage attach/detach storage operation. Currently that object contains reference to an associated PV when the driver is not used inline.  However,  it must be extended to support information from `CSIVolumeSource` (see above) when an inline volume is being attached.
+
+```go
+// VolumeAttachmentSpec is the specification of a VolumeAttachment request.
+type VolumeAttachmentSpec struct {
+    // <snip>
+
+	// Source represents the volume that should be attached.
+	Source VolumeAttachmentSource
+}
+
+// VolumeAttachmentSource represents a volume that should be attached, either
+// PersistentVolume or a volume in-lined in a Pod.
+// Exactly one member can be set.
+type VolumeAttachmentSource struct {
+	// Name of the persistent volume to attach.
+	// +optional
+	PersistentVolumeName *string
+
+	// InlineVolumeSource represents the source location of a in-line volume in a pod to attach.
+	// +optional
+    InlineVolumeSource *InlineVolumeSource
+}
+
+// InlineVolumeSource represents the source location of a in-line volume in a pod.
+type InlineVolumeSource struct {
+	// VolumeSource is copied from the pod. It ensures that attacher has enough
+	// information to detach a volume when the pod is deleted before detaching.
+	// Only CSIVolumeSource can be set.
+	// Required.
+	CSIVolumeSource v1.VolumeSource
+
+	// Namespace of the pod with in-line volume. It is used to resolve
+	// references to Secrets in VolumeSource.
+	// Required.
+	Namespace string
+}
+```
+The following steps outline how the API is used to track attachment and detachment of volumes:
+* The external CSI A/D controller **copies whole `VolumeSource`**  from `Pod` into `VolumeAttachment`. This allows external CSI attacher to detach volumes for deleted pods without keeping any internal database of attached VolumeSources.
+* Validation of `VolumeSource` will ensure that only `CSIVolumeSource` is being copied.  
+* External CSI attacher must be extended to  process either `PersistentVolumeName` or `VolumeSource`.
+* **External attacher may need permission to access secrets in any pod namespace** where inline volume is specified.
+* CSI `ControllerUnpublishVolume` call (~ volume detach) will require the secrets to be available at detach time. 
+* If user deletes secrets in a pod's namespace that was used to attach an inline volume, the external attacher will fail during detach (volume will remain attached), reporting errors about missing secrets to user.
```

Comment:
> reporting errors about missing secrets to user

using what mechanism?

Reply:
pod events?

Reply:
@liggitt this will have to be done with events and logging.

Reply:
if there's an error attaching a volume, is it reported as an event into all namespaces that have pods waiting for that attachment, or just into the namespace referenced by the VolumeAttachment object?